composer/datasets/streaming/dataset.py

```diff
@@ -0,0 +1,379 @@
+            shard_min_id = self.index.shard_begins[shard]
+            shard_max_id = self.index.shard_ends[shard] - 1
+            min_id = max(part_min_id, shard_min_id)
+            max_id = min(part_max_id, shard_max_id)
+            new_ids += list(range(min_id, max_id + 1))
+
+        with self._lock:
+            # Extend and optionally reshuffle the remaining samples of any
+            # epochs we have in progress.
+            if self.shuffle:
+                if not self._are_all_shards_downloaded:
+                    self._downloaded_ids.extend(new_ids)
+                    np.random.shuffle(self._downloaded_ids)
+                for todo_ids in self._epoch_to_todo_ids.values():
+                    todo_ids.extend(new_ids)
+                    np.random.shuffle(todo_ids)
+            else:
+                if not self._are_all_shards_downloaded:
+                    self._downloaded_ids.extend(new_ids)
+                for todo_ids in self._epoch_to_todo_ids.values():
+                    todo_ids.extend(new_ids)
+
+    def _load_shards_if_downloaded(self, shards: Sequence[int], part_min_id: int, part_max_id: int) -> List[int]:
+        """Load any of the given shards that are already present in the cache, returning the missing shards.
+
+        Args:
+            shards (Sequence[int]): The shards to attempt to load.
+            part_min_id (int): Minimum sample ID of this partition.
+            part_max_id (int): Maximum sample ID of this partition.
+
+        Returns:
+            list of int: The shards that remain to be loaded.
+        """
+        downloaded = []
+        missing = []
+        for shard in sorted(shards):
+            basename = get_shard_basename(shard)
+            local = os.path.join(self.local, basename)
+            if os.path.exists(local):
+                downloaded.append(shard)
+            else:
+                missing.append(shard)
+        if downloaded:
+            self._load_shards(downloaded, part_min_id, part_max_id)
+        return missing
+
+    def _done_loading(self) -> None:
+        """Callback on completion of loading my shards."""
+        with self._lock:
+            self._are_all_shards_downloaded = True
+
+    def _download_thread(self, shards: Sequence[int], part_min_id: int, part_max_id: int) -> None:
+        """Background thread to download and assimilate missing shards.
+
+        Args:
+            shards (list of int): The shards remaining to be downloaded.
+            part_min_id (int): Minimum sample ID of this partition.
+            part_max_id (int): Maximum sample ID of this partition.
+        """
+        shards = list(shards)
+        if self.shuffle:
+            np.random.shuffle(shards)
+        for shard in shards:
+            basename = get_shard_basename(shard)
+            self._download_if_missing(basename)
+            shards = shard,
+            self._load_shards(shards, part_min_id, part_max_id)
+        self._done_loading()
```

Comment:
Will this thread gracefully handle intermittent network errors? In addition, if it does crash, will it also crash the training process?